File: src/gui/src/joint_state_publisher.cpp
#include "gui/joint_state_publisher.hpp"
#include <chrono>
#include <functional>

using namespace std::chrono_literals;

namespace gui {

JointStatePublisher::JointStatePublisher(const rclcpp::NodeOptions & options)
: Node("joint_state_publisher", options)
{
  // Declare and get parameters
  this->declare_parameter("publish_rate", 10.0);
  this->declare_parameter("joint_names", std::vector<std::string>{"joint1", "joint2", "joint3", "joint4", "joint5", "gripper"});
  
  publish_rate_ = this->get_parameter("publish_rate").as_double();
  joint_names_ = this->get_parameter("joint_names").as_string_array();
  
  // Initialize joint states
  const size_t num_joints = joint_names_.size();
  joint_positions_.resize(num_joints, 0.0);
  joint_velocities_.resize(num_joints, 0.0);
  joint_efforts_.resize(num_joints, 0.0);
  
  // Create publisher
  joint_state_pub_ = this->create_publisher<sensor_msgs::msg::JointState>(
    "joint_states", 10);
    
  // Create mock service client
  joint_control_client_ = std::make_shared<mock::JointControlClient>();
    
  // Create timer for publishing joint states
  auto timer_period = std::chrono::duration<double>(1.0 / publish_rate_);
  timer_ = this->create_wall_timer(
    timer_period, std::bind(&JointStatePublisher::timerCallback, this));
    
  RCLCPP_INFO(this->get_logger(), "Joint State Publisher initialized with %zu joints", num_joints);
}

JointStatePublisher::~JointStatePublisher()
{
  // Clean up resources if needed
}

void JointStatePublisher::timerCallback()
{
  // Update joint states from actual robot
  updateJointStates();
  
  // Create and publish joint state message
  auto joint_state_msg = std::make_unique<sensor_msgs::msg::JointState>();
  
  // Set header
  joint_state_msg->header.stamp = this->now();
  
  // Lock mutex for thread safety
  std::lock_guard<std::mutex> lock(joint_state_mutex_);
  
  // Set joint states
  joint_state_msg->name = joint_names_;
  joint_state_msg->position = joint_positions_;
  joint_state_msg->velocity = joint_velocities_;
  joint_state_msg->effort = joint_efforts_;
  
  // Publish message
  joint_state_pub_->publish(std::move(joint_state_msg));
}

void JointStatePublisher::updateJointStates()
{
  // This method would query the actual robot for joint states
  // For now, we'll just use placeholder values
  
  // Lock mutex for thread safety
  std::lock_guard<std::mutex> lock(joint_state_mutex_);
  
  // Update with placeholder values (would be replaced with actual robot data)
  for (size_t i = 0; i < joint_positions_.size(); ++i) {
    // Simulate some movement
    joint_positions_[i] = 0.1 * sin(this->now().seconds() + i);
    joint_velocities_[i] = 0.1 * cos(this->now().seconds() + i);
    joint_efforts_[i] = 0.0;
  }
}

} // namespace gui

#include "rclcpp_components/register_node_macro.hpp"
RCLCPP_COMPONENTS_REGISTER_NODE(gui::JointStatePublisher) 

File: src/gui/src/gui_python/joint_state_publisher.py
#!/usr/bin/env python3

"""
关节状态发布器模块。
用于发布机器人关节状态。
"""

import threading
import math
import time
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import JointState
from std_msgs.msg import Header


class JointStatePublisherNode(Node):
    """
    关节状态发布器节点类。
    负责发布机器人关节状态信息。
    """
    
    def __init__(self):
        """初始化节点。"""
        super().__init__('joint_state_publisher')
        
        # 声明并获取参数
        self.declare_parameter('publish_rate', 10.0)
        self.declare_parameter('joint_names', 
                             ['joint1', 'joint2', 'joint3', 'joint4', 'joint5', 'gripper'])
        
        self.publish_rate = self.get_parameter('publish_rate').get_parameter_value().double_value
        self.joint_names = self.get_parameter('joint_names').get_parameter_value().string_array_value
        
        # 如果joint_names参数为空，使用默认值
        if not self.joint_names:
            self.joint_names = ['joint1', 'joint2', 'joint3', 'joint4', 'joint5', 'gripper']
        
        # 初始化关节状态
        self.num_joints = len(self.joint_names)
        self.joint_positions = [0.0] * self.num_joints
        self.joint_velocities = [0.0] * self.num_joints
        self.joint_efforts = [0.0] * self.num_joints
        
        # 创建发布器
        self.joint_state_pub = self.create_publisher(
            JointState,
            'joint_states',
            10)
        
        # 创建定时器
        timer_period = 1.0 / self.publish_rate  # 秒
        self.timer = self.create_timer(timer_period, self.timer_callback)
        
        # 创建锁以确保线程安全
        self._lock = threading.Lock()
        
        self.get_logger().info(f'Joint State Publisher initialized with {self.num_joints} joints')
    
    def timer_callback(self):
        """定时器回调函数，定期发布关节状态。"""
        # 更新关节状态
        self.update_joint_states()
        
        # 创建并发布关节状态消息
        joint_state_msg = JointState()
        
        # 设置头
        joint_state_msg.header = Header()
        joint_state_msg.header.stamp = self.get_clock().now().to_msg()
        
        # 获取锁以确保线程安全
        with self._lock:
            # 设置关节状态
            joint_state_msg.name = self.joint_names
            joint_state_msg.position = self.joint_positions
            joint_state_msg.velocity = self.joint_velocities
            joint_state_msg.effort = self.joint_efforts
        
        # 发布消息
        self.joint_state_pub.publish(joint_state_msg)
    
    def update_joint_states(self):
        """更新关节状态，现在使用模拟数据。"""
        # 这个方法会查询实际机器人的关节状态
        # 现在我们只使用占位符值
        
        # 获取锁以确保线程安全
        with self._lock:
            # 使用占位符值更新（将替换为实际机器人数据）
            current_time = time.time()
            for i in range(len(self.joint_positions)):
                # 模拟一些运动
                self.joint_positions[i] = 0.1 * math.sin(current_time + i)
                self.joint_velocities[i] = 0.1 * math.cos(current_time + i)
                self.joint_efforts[i] = 0.0 